src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs

```diff
@@ -0,0 +1,214 @@
+using System.Threading.Tasks;
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     Responsible for listening to project loaded events and firing <see cref="IUnconfiguredProjectTasksService.PrioritizedProjectLoadedInHost"/> and 
+    ///     <see cref="IUnconfiguredProjectTasksService.ProjectLoadedInHost"/>.
+    /// </summary>
+    [Export(typeof(ILoadedInHostListener))]
+    internal class VsSolutionEventListener : OnceInitializedOnceDisposedAsync, IVsSolutionEvents, IVsPrioritizedSolutionEvents, ILoadedInHostListener
+    {
+        private readonly IVsService<IVsSolution> _solution;
+        private readonly IProjectThreadingService _threadingService;
+        private uint _cookie = VSConstants.VSCOOKIE_NIL;
+
+        [ImportingConstructor]
+        public VsSolutionEventListener(IVsService<SVsSolution, IVsSolution> solution, IProjectThreadingService threadingService)
+            : base(threadingService.JoinableTaskContext)
+        {
+            _solution = solution;
+            _threadingService = threadingService;
+        }
+
+        public Task StartListeningAsync()
+        {
+            return InitializeAsync(CancellationToken.None);
+        }
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            await _threadingService.SwitchToUIThread();
+
+            HResult result = _solution.Value.AdviseSolutionEvents(this, out _cookie);
+            if (result.Failed)
+                throw result.Exception;
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            await _threadingService.SwitchToUIThread();
+
+            if (_cookie != VSConstants.VSCOOKIE_NIL)
+            {
+                // Due to synchronous disposable, we should be on UI thread,
+                // accessing IVsService<T>.Value will enforce that.
+                HResult result = _solution.Value.UnadviseSolutionEvents(_cookie);
+                if (result.Failed)
+                    throw result.Exception;
+
+                _cookie = VSConstants.VSCOOKIE_NIL;
+            }
+        }
+
+        public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
+        {
+            UnconfiguredProjectTasksService tasksService = GetUnconfiguredProjectTasksServiceIfApplicable(pHierarchy);
+            if (tasksService != null)
+            {
+                tasksService.OnProjectLoadedInHost();
+            }
+
+            return HResult.OK;
+        }
+
+        public int PrioritizedOnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
+        {
+            UnconfiguredProjectTasksService tasksService = GetUnconfiguredProjectTasksServiceIfApplicable(pHierarchy);
+            if (tasksService != null)
+            {
+                tasksService.OnPrioritizedProjectLoadedInHost();
```

Comment:
Consider: `taskService?.OnPrioritizedProjectLoadedInHost();` (and above)

Reply:
Done.